### PR TITLE
Loss functions

### PR DESCRIPTION
**Pull request by ramasesh**
Added loss function for multi-class classification

### CODE REVIEW

File: renn/utils.py [hunk 1]
```diff
@@ -137,3 +139,85 @@ def one_hot(labels, num_classes, dtype=jnp.float32):
...
+
+  return selected_outputs
+
+def make_loss_function(network_apply_fun, loss_type = 'xent', num_outputs = 1):
+  """ Given a network function, return a loss function which maps a tuple of
+  network parameters and a training batch to a loss value. """
+
+  ALLOWED_LOSSES = ['xent', 'mse']
+  assert loss_type in ALLOWED_LOSSES
+
```
Comment (nirum) [inline renn/utils.py:176]: easier to just catch this in an `else` branch below

File: renn/utils.py [hunk 2]
```diff
@@ -137,3 +139,85 @@ def one_hot(labels, num_classes, dtype=jnp.float32):
...
+  # select element
+  selected_outputs = jnp.take_along_axis(sequences, indices_shaped, axis=1)
+
+  # remove sequence dimension
+  selected_outputs = jnp.squeeze(selected_outputs, axis=1)
+
+  return selected_outputs
+
+def make_loss_function(network_apply_fun, loss_type = 'xent', num_outputs = 1):
+  """ Given a network function, return a loss function which maps a tuple of
```
Comment (nirum) [inline renn/utils.py:171]: Suggestion for a cleaner API: just have the loss function be an argument that is directly passed as an argument.

That simplifies the number of arguments this function needs (instead of having both a `loss_type` and a `num_outputs`, it just has a single argument, the function). In addition, as we add potentially new loss functions (such as MSE) we do not need to edit/update this function.

File: renn/utils.py [hunk 3]
```diff
@@ -137,3 +139,85 @@ def one_hot(labels, num_classes, dtype=jnp.float32):
     one_hot_labels: array with shape (num_examples, num_classes).
   """
   return jnp.array(jnp.array(labels)[:, None] == jnp.arange(num_classes), dtype)
+
+
+def select_output(sequences, indices):
```
Comment (nirum) [inline renn/utils.py:144]: Maybe just call this `select`, since it can be applied to anything (not just the output)?

We should also add some simple tests for this function ... I find that this kind of array manipulation / filtering code benefits the most from simple unit tests.

Comment (ramasesh) [reply renn/utils.py:144]: Agree, I called it select and renamed the variable `selected_outputs` inside the function to `selected_elements`.  

For the test, I'll open an issue.

File: renn/utils.py (hunk 1 above)
Comment (ramasesh) [reply renn/utils.py:176]: Fixed suggestion below, makes this a non-issue

File: renn/utils.py (hunk 2 above)
Comment (ramasesh) [reply renn/utils.py:171]: Agree, made this change
